ArgumentOutOfRangeException in Android shell when page becomes invisible

### Description

I have an app with dynamic shell content - pages become visible or invisible depending on the state of another system, via data binding.
This was fine in earlier versions, but in 6.0.300-rc.3 (which I assume is the version shipped with VS 2022 17.3.0 preview 1?) an exception is thrown when an active page becomes invisible.

### Steps to Reproduce

(Code is shown in more detail below)

1. Create a new MAUI app called ShellPageRemoved (to fit in with the code below)
2. Add a new file SimpleViewModel (content below) with a `Visible` property
3. Add a new ContentPage called Page2, containing a button that changes `SimpleViewModel.Visible`
4. Add Page2 to the shell, with a visibility based on `SimpleViewModel.Visible`
5. Set the `BindingContext` of the app to a `SimpleViewModel`
6. Start the app in the Android emulator or a real device (in the debugger)
7. Navigate to Page 2
8. Click the button to make the page disappear
9. Observe the exception

Code:

SimpleViewModel.cs:
```csharp
using System.ComponentModel;

namespace ShellPageRemoved;

public class SimpleViewModel : INotifyPropertyChanged
{
    public event PropertyChangedEventHandler PropertyChanged;

    public bool Visible { get; set; } = true;

    public void SetVisibleFalse()
    {
        Visible = false;
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(Visible)));
    }
}
```

Page2.xaml:
```xml
<?xml version="1.0" encoding="utf-8" ?>
<ContentPage xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             x:Class="ShellPageRemoved.Page2"
             Title="Page2">
    <StackLayout>
        <Button Text="Click me to make me disappear" Clicked="SetVisibleFalse" />
    </StackLayout>
</ContentPage>
```

Page.xaml.cs:
```csharp
namespace ShellPageRemoved;

public partial class Page2 : ContentPage
{
	public Page2()
	{
		InitializeComponent();
	}

    private void SetVisibleFalse(object sender, EventArgs e)
    {
		SimpleViewModel vm = (SimpleViewModel)BindingContext;
		vm.SetVisibleFalse();
    }
}
```

App.xaml.cs change (all other lines remain the same):

```csharp
MainPage = new AppShell { BindingContext = new SimpleViewModel() };
```

AppShell.xaml:
```xml
<?xml version="1.0" encoding="UTF-8" ?>
<Shell
    x:Class="ShellPageRemoved.AppShell"
    xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
    xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
    xmlns:local="clr-namespace:ShellPageRemoved"
    Shell.FlyoutBehavior="Disabled">

    <TabBar>
        <ShellContent Title="Home" ContentTemplate="{DataTemplate local:MainPage}" />
        <ShellContent Title="Page 2" ContentTemplate="{DataTemplate local:Page2}" IsVisible="{Binding Visible}"/>
    </TabBar>
</Shell>
```

### Version with bug

Release Candidate 3 (current)

### Last version that worked well

Release Candidate 2

### Affected platforms

Android

### Affected platform versions

Only tried with "Android 11.0 - API 30" emulator and a Pixel 6

### Did you find any workaround?

Not yet. I *suspect* that if I remove the databinding from visibility and navigate away from the page before changing the visibility, it will be okay.

### Relevant log output

```shell
System.ArgumentOutOfRangeException
  Message=Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index')
[mono-rt] [ERROR] FATAL UNHANDLED EXCEPTION: System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index')
[mono-rt]    at System.Collections.Generic.List`1[[Microsoft.Maui.Controls.ShellContent, Microsoft.Maui.Controls, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].get_Item(Int32 index)
[mono-rt]    at System.Collections.ObjectModel.Collection`1[[Microsoft.Maui.Controls.ShellContent, Microsoft.Maui.Controls, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].get_Item(Int32 index)
[mono-rt]    at System.Collections.ObjectModel.ReadOnlyCollection`1[[Microsoft.Maui.Controls.ShellContent, Microsoft.Maui.Controls, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].get_Item(Int32 index)
[mono-rt]    at Microsoft.Maui.Controls.Platform.Compatibility.ShellSectionRenderer.OnPageSelected(Int32 position) in D:\a\_work\1\s\src\Controls\src\Core\Compatibility\Handlers\Shell\Android\ShellSectionRenderer.cs:line 291
[mono-rt]    at Microsoft.Maui.Controls.Platform.Compatibility.ShellSectionRenderer.ViewPagerPageChanged.OnPageSelected(Int32 position) in D:\a\_work\1\s\src\Controls\src\Core\Compatibility\Handlers\Shell\Android\ShellSectionRenderer.cs:line 338
[mono-rt]    at AndroidX.ViewPager2.Widget.ViewPager2.OnPageChangeCallback.n_OnPageSelected_I(IntPtr jnienv, IntPtr native__this, Int32 position) in C:\a\_work\1\s\generated\androidx.viewpager2.viewpager2\obj\Release
et6.0-android\generated\src\AndroidX.ViewPager2.Widget.ViewPager2.cs:line 318
[mono-rt]    at Android.Runtime.JNINativeWrapper.Wrap_JniMarshal_PPI_V(_JniMarshal_PPI_V callback, IntPtr jnienv, IntPtr klazz, Int32 p0) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Android.Runtime/JNINativeWrapper.g.cs:line 55
```

--- a/src/Controls/src/Core/Compatibility/Handlers/Shell/Android/ShellFragmentStateAdapter.cs
+++ b/src/Controls/src/Core/Compatibility/Handlers/Shell/Android/ShellFragmentStateAdapter.cs
@@ -27,11 +27,10 @@ namespace Microsoft.Maui.Controls.Platform.Compatibility
 		{
 			_mauiContext = mauiContext;
 			_shellSection = shellSection;
-			SectionController.ItemsCollectionChanged += OnItemsCollectionChanged;
 			_items = SectionController.GetItems();
 		}
 
-		protected virtual void OnItemsCollectionChanged(object sender, NotifyCollectionChangedEventArgs e)
+		public void OnItemsCollectionChanged(object sender, NotifyCollectionChangedEventArgs e)
 		{
 			_items = SectionController.GetItems();
 			var removeList = new List<long>();
@@ -42,8 +41,6 @@ namespace Microsoft.Maui.Controls.Platform.Compatibility
 
 			foreach (var remove in removeList)
 				_createdShellContent.Remove(remove);
-
-			NotifyDataSetChanged();
 		}
 
 		public int CountOverride { get; set; }
@@ -70,6 +67,14 @@ namespace Microsoft.Maui.Controls.Platform.Compatibility
 
 		public override bool ContainsItem(long itemId)
 		{
+			if (_createdShellContent.TryGetValue(itemId, out var shellContent) &&
+				!_items.Contains(shellContent))
+			{
+				// This means a data set change was triggered but the INCC change hasn't
+				// propagated from our xplat code to here yet
+				_createdShellContent.Remove(itemId);
+			}
+
 			return _createdShellContent.ContainsKey(itemId);
 		}
 
@@ -82,7 +87,6 @@ namespace Microsoft.Maui.Controls.Platform.Compatibility
 
 			if (disposing)
 			{
-				SectionController.ItemsCollectionChanged -= OnItemsCollectionChanged;
 				_shellSection = null;
 
 				_items = null;

--- a/src/Controls/src/Core/Compatibility/Handlers/Shell/Android/ShellSectionRenderer.cs
+++ b/src/Controls/src/Core/Compatibility/Handlers/Shell/Android/ShellSectionRenderer.cs
@@ -12,6 +12,7 @@ using AndroidX.Fragment.App;
 using AndroidX.ViewPager.Widget;
 using AndroidX.ViewPager2.Widget;
 using Google.Android.Material.Tabs;
+using Microsoft.Extensions.Logging;
 using AToolbar = AndroidX.AppCompat.Widget.Toolbar;
 using AView = Android.Views.View;
 
@@ -234,10 +235,23 @@ namespace Microsoft.Maui.Controls.Platform.Compatibility
 			}
 		}
 
-		void SafeNotifyDataSetChanged()
+		void SafeNotifyDataSetChanged(int iteration = 0)
 		{
 			if (_disposed)
 				return;
+
+			if (!_viewPager.IsAlive())
+				return;
+
+			if (iteration >= 10)
+			{
+				// It's very unlikely this will happen but just in case there's a scenario
+				// where we might hit an infinite loop we're adding an exit strategy
+				MauiContext.CreateLogger<ShellSectionRenderer>()
+					.LogWarning("ViewPager2 stuck in layout, unable to NotifyDataSetChanged;");
+
+				return;
+			}
 
 			if (_viewPager?.Adapter is ShellFragmentStateAdapter adapter)
 			{
@@ -249,7 +263,7 @@ namespace Microsoft.Maui.Controls.Platform.Compatibility
 				}
 				else
 				{
-					_viewPager.Post(SafeNotifyDataSetChanged);
+					_viewPager.Post(() => SafeNotifyDataSetChanged(++iteration));
 				}
 			}
 		}
@@ -330,7 +344,8 @@ namespace Microsoft.Maui.Controls.Platform.Compatibility
 			var visibleItems = SectionController.GetItems();
 
 			// This mainly happens if all of the items that are part of this shell section 
-			// vanish. Android calls `OnPageSelected` with the last item left in the list
+			// vanish. Android calls `OnPageSelected` with position zero even though the view pager is
+			// empty
 			if (visibleItems.Count >= position)
 				return;
 

--- a/src/Controls/src/Core/Compatibility/Handlers/Shell/Android/ShellSectionRenderer.cs
+++ b/src/Controls/src/Core/Compatibility/Handlers/Shell/Android/ShellSectionRenderer.cs
@@ -226,6 +226,32 @@ namespace Microsoft.Maui.Controls.Platform.Compatibility
 		protected virtual void OnItemsCollectionChanged(object sender, NotifyCollectionChangedEventArgs e)
 		{
 			UpdateTablayoutVisibility();
+
+			if (_viewPager?.Adapter is ShellFragmentStateAdapter adapter)
+			{
+				adapter.OnItemsCollectionChanged(sender, e);
+				SafeNotifyDataSetChanged();
+			}
+		}
+
+		void SafeNotifyDataSetChanged()
+		{
+			if (_disposed)
+				return;
+
+			if (_viewPager?.Adapter is ShellFragmentStateAdapter adapter)
+			{
+				// https://stackoverflow.com/questions/43221847/cannot-call-this-method-while-recyclerview-is-computing-a-layout-or-scrolling-wh
+				// ViewPager2 is based on RecyclerView which really doesn't like NotifyDataSetChanged when a layout is happening
+				if (!_viewPager.IsInLayout)
+				{
+					adapter.NotifyDataSetChanged();
+				}
+				else
+				{
+					_viewPager.Post(SafeNotifyDataSetChanged);
+				}
+			}
 		}
 
 		void UpdateTablayoutVisibility()
@@ -258,7 +284,9 @@ namespace Microsoft.Maui.Controls.Platform.Compatibility
 				var newIndex = SectionController.GetItems().IndexOf(ShellSection.CurrentItem);
 
 				if (SectionController.GetItems().Count != _viewPager.ChildCount)
-					_viewPager.Adapter.NotifyDataSetChanged();
+				{
+					SafeNotifyDataSetChanged();
+				}
 
 				if (newIndex >= 0)
 				{
@@ -299,7 +327,14 @@ namespace Microsoft.Maui.Controls.Platform.Compatibility
 				return;
 
 			var shellSection = ShellSection;
-			var shellContent = SectionController.GetItems()[position];
+			var visibleItems = SectionController.GetItems();
+
+			// This mainly happens if all of the items that are part of this shell section 
+			// vanish. Android calls `OnPageSelected` with the last item left in the list
+			if (visibleItems.Count >= position)
+				return;
+
+			var shellContent = visibleItems[position];
 
 			if (shellContent == shellSection.CurrentItem)
 				return;
@@ -314,7 +349,7 @@ namespace Microsoft.Maui.Controls.Platform.Compatibility
 			}
 			else if (shellSection?.CurrentItem != null)
 			{
-				var currentPosition = SectionController.GetItems().IndexOf(shellSection.CurrentItem);
+				var currentPosition = visibleItems.IndexOf(shellSection.CurrentItem);
 				_selecting = true;
 
 				// Android doesn't really appreciate you calling SetCurrentItem inside a OnPageSelected callback.

--- a/src/Controls/src/Core/HandlerImpl/Page.Impl.cs
+++ b/src/Controls/src/Core/HandlerImpl/Page.Impl.cs
@@ -7,6 +7,8 @@ namespace Microsoft.Maui.Controls
 	/// <include file="../../../docs/Microsoft.Maui.Controls/Page.xml" path="Type[@FullName='Microsoft.Maui.Controls.Page']/Docs" />
 	public partial class Page : IView, ITitledElement, IToolbarElement
 	{
+		internal bool HasNavigatedTo { get; private set; }
+
 		Paint IView.Background
 		{
 			get
@@ -40,6 +42,7 @@ namespace Microsoft.Maui.Controls
 
 		internal void SendNavigatedTo(NavigatedToEventArgs args)
 		{
+			HasNavigatedTo = true;
 			NavigatedTo?.Invoke(this, args);
 			OnNavigatedTo(args);
 		}
@@ -52,6 +55,7 @@ namespace Microsoft.Maui.Controls
 
 		internal void SendNavigatedFrom(NavigatedFromEventArgs args)
 		{
+			HasNavigatedTo = false;
 			NavigatedFrom?.Invoke(this, args);
 			OnNavigatedFrom(args);
 		}

--- a/src/Controls/src/Core/Shell/ShellContent.cs
+++ b/src/Controls/src/Core/Shell/ShellContent.cs
@@ -209,6 +209,11 @@ namespace Microsoft.Maui.Controls
 
 		public static implicit operator ShellContent(TemplatedPage page)
 		{
+			if (page.Parent != null)
+			{
+				return (ShellContent)page.Parent;
+			}
+
 			var shellContent = new ShellContent();
 
 			var pageRoute = Routing.GetRoute(page);

--- a/src/Controls/tests/Core.UnitTests/ShellTests.cs
+++ b/src/Controls/tests/Core.UnitTests/ShellTests.cs
@@ -72,6 +72,34 @@ namespace Microsoft.Maui.Controls.Core.UnitTests
 
 
 		[Test]
+		public void SettingCurrentItemOnShellViaContentPage()
+		{
+			var page1 = new ContentPage();
+			var page2 = new ContentPage();
+			var shell = new TestShell()
+			{
+				Items =
+				{
+					new TabBar()
+					{
+						Items =
+						{
+							new ShellContent() { Content = page1 },
+							new ShellContent() { Content = page2 },
+						}
+					}
+				}
+			};
+
+			shell.CurrentItem = page2;
+			Assert.AreEqual(1, shell.Items.Count);
+			Assert.AreEqual(2, shell.Items[0].Items.Count);
+			Assert.AreEqual(1, shell.Items[0].Items[0].Items.Count);
+			Assert.AreEqual(1, shell.Items[0].Items[1].Items.Count);
+			Assert.AreEqual(shell.CurrentItem.CurrentItem, shell.Items[0].Items[1]);
+		}
+
+		[Test]
 		public void SetCurrentItemAddsToShellCollection()
 		{
 			var shell = new Shell();

--- a/src/Controls/tests/DeviceTests/Elements/Shell/ShellTests.cs
+++ b/src/Controls/tests/DeviceTests/Elements/Shell/ShellTests.cs
@@ -190,6 +190,45 @@ namespace Microsoft.Maui.DeviceTests
 			});
 		}
 
+
+
+		[Fact(DisplayName = "Correctly Adjust to Making Currently Visible Shell Page Invisible")]
+		public async Task CorrectlyAdjustToMakingCurrentlyVisibleShellPageInvisible()
+		{
+			SetupBuilder();
+
+			var page1 = new ContentPage();
+			var page2 = new ContentPage();
+
+			var shell = await CreateShellAsync((shell) =>
+			{
+				var tabBar = new TabBar()
+				{
+					Items =
+					{
+						new ShellContent(){ Content = page1 },
+						new ShellContent(){ Content = page2 },
+					}
+				};
+
+				shell.Items.Add(tabBar);
+			});
+
+			await CreateHandlerAndAddToWindow<ShellHandler>(shell, async (handler) =>
+			{
+				await OnNavigatedToAsync(page1);
+				shell.CurrentItem = page2;
+				await OnNavigatedToAsync(page2);
+				page2.IsVisible = false;
+				await OnNavigatedToAsync(page1);
+				Assert.Equal(shell.CurrentPage, page1);
+				page2.IsVisible = true;
+				shell.CurrentItem = page2;
+				await OnNavigatedToAsync(page2);
+				Assert.Equal(shell.CurrentPage, page2);
+			});
+		}
+
 		[Fact(DisplayName = "Empty Shell")]
 		public async Task DetailsViewUpdates()
 		{

--- a/src/Controls/tests/DeviceTests/HandlerTestBase.cs
+++ b/src/Controls/tests/DeviceTests/HandlerTestBase.cs
@@ -290,21 +290,25 @@ namespace Microsoft.Maui.DeviceTests
 			return taskCompletionSource.Task.WaitAsync(timeOut.Value);
 		}
 
-		protected Task OnNavigatedToAsync(Page page, TimeSpan? timeOut = null)
+		protected async Task OnNavigatedToAsync(Page page, TimeSpan? timeOut = null)
 		{
+			await OnLoadedAsync(page, timeOut);
+
+			if (page.HasNavigatedTo)
+				return;
+
 			timeOut = timeOut ?? TimeSpan.FromSeconds(2);
 			TaskCompletionSource<object> taskCompletionSource = new TaskCompletionSource<object>();
 
 			page.NavigatedTo += NavigatedTo;
 
-			return taskCompletionSource.Task.WaitAsync(timeOut.Value);
+			await taskCompletionSource.Task.WaitAsync(timeOut.Value);
 			void NavigatedTo(object sender, NavigatedToEventArgs e)
 			{
 				taskCompletionSource.SetResult(true);
 				page.NavigatedTo -= NavigatedTo;
 			}
 		}
-
 
 		protected Task OnFrameSetToNotEmpty(VisualElement frameworkElement, TimeSpan? timeOut = null)
 		{

